In [2]:
import sys
!{sys.executable} -m pip install pysam
import pysam

aligned_file=pysam.AlignmentFile('merged-tumor.bam')
num_unmapped=0
num_zero_qual=0
num_reads=0
average_mapq=0
average_mapq_nonzero=0

for i,read in enumerate(aligned_file):
    num_reads+=1
    if i==0:
        print("\nFirst read:\n",read)
        print("\nFirst read parameters values: \n")
        print("FLAG: ",read.flag)
        print("RNAME: ",read.reference_id)
        print("POS: ",read.reference_start)
        print("Mapping quality: ",read.mapping_quality)
        print("CIGAR: ",read.cigarstring)
        print("MRNM: ",read.next_reference_id)
        print("MPOS: ",read.next_reference_start)
        print("ISIZE: ",read.template_length)
        print("SEQ: ",read.query_sequence)
        print("QUAL: ",read.query_qualities)
            
    if i%400000==0 and i>399999:
        print("FLAG(read num."+str(i)+"): ",read.flag)
         
    average_mapq+=read.mapping_quality
    if read.is_unmapped:
        num_unmapped+=1
   
    if read.mapping_quality==0:
        num_zero_qual+=1
    else:
        average_mapq_nonzero+=read.mapping_quality
    
    
average_mapq/=num_reads   
average_mapq_nonzero/=(num_reads-num_zero_qual)


print("\nTotal number of unmapped reads: ",num_unmapped)
print("Total number of reads: ",num_reads)
print("Total number of reads with mapping quality equal to 0: ",num_zero_qual)
print("Average mapping quality for all the reads:",average_mapq)
print("Average mapping quality for reads with mapping quality not equal to 0:",average_mapq_nonzero)





First read:
 C0HVYACXX120402:7:1207:5722:57044	1187	20	9483248	27	76M	20	9483381	76	TTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAAAAAAAAACCAGAATGTGCCATTCTGATTTAAACTG	array('B', [28, 28, 27, 29, 31, 30, 31, 31, 29, 31, 35, 30, 29, 31, 34, 30, 29, 23, 41, 32, 20, 30, 29, 34, 34, 29, 30, 31, 30, 30, 30, 33, 33, 26, 39, 12, 25, 19, 32, 30, 35, 28, 35, 33, 23, 33, 35, 36, 30, 38, 33, 41, 34, 35, 31, 33, 23, 30, 30, 36, 27, 32, 29, 34, 35, 41, 33, 31, 33, 29, 32, 32, 31, 31, 31, 34])	[('XA', 'GL000217.1,-110754,76M,1;'), ('BD', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('MD', '76'), ('RG', '1'), ('BI', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('NM', 0), ('MQ', 27), ('AS', 76), ('XS', 71)]

First read parameters values: 

FLAG:  1187
RNAME:  20
POS:  9483248
Mapping quality:  27
CIGAR:  76M
MRNM:  20
MPOS:  9483381
ISIZE:  209
SEQ:  TTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAAAAAAAAACCAGAATGTGCCATTCTGATTTAAACTG
QUAL:  arr

First read parameters inspection:<br>
1.FLAG=1187<br>
  -read paired (0x1)<br>
  -read mapped in proper pair (0x2)<br>
  -mate reverse strand (0x20)<br>
  -second in pair (0x80)<br>
  -read is PCR or optical duplicate (0x400)<br>
  
2.RNAME=20 (Read name)

3.POS=9483248 (Position of read alignment)

4.Mapping quality=27

5.CIGAR=76M (76 nucleotides are matched or mismatched)

6.MRNM=20 (Name of mate)

7.MPOS=9483381(Position of mate)


8.ISIZE=209 (Template length)


9.SEQ=TTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAAAAAAAAACCAGAATGTGCCATTCTGATTTAAACTG (Sequence of nucleotides in this read)

10.QUAL=array('B',[28, 28, 27, 29, 31, 30, 31, 31, 29, 31, 35, 30, 29, 31, 34, 30, 29, 23, 41, 32, 20, 30, 29, 34, 34, 29, 30, 31, 30, 30, 30, 33, 33, 26, 39, 12, 25, 19, 32, 30, 35, 28, 35, 33, 23, 33, 35, 36, 30, 38, 33, 41, 34, 35, 31, 33, 23, 30, 30, 36, 27, 32, 29, 34, 35, 41, 33, 31, 33, 29, 32, 32, 31, 31, 31, 34])
(Quality scores and its not ASCII-encoded value as it is used in BAM files so there is no offset 33 that needs to be substracted)

FLAG inspection:

Read number 400000 and 800000:<br>
FLAG=83<br>
-read paired (0x1)<br>
-read mapped in proper pair (0x2)<br>
-read reverse strand (0x10)<br>
-first in pair (0x40)<br>

Read number 1200000 and 1600000:<br>
FLAG=163<br>
-read paired (0x1)<br>
-read mapped in proper pair (0x2)<br>
-mate reverse strand (0x20)<br>
-second in pair (0x80)<br>

Read number 2400000:<br>
FLAG=121<br>
-read paired (0x1)<br>
-mate unmapped (0x8)<br>
-read reverse strand (0x10)<br>
-mate reverse strand (0x20)<br>
-first in pair (0x40)<br>

Read number 2800000:<br>
FLAG=99<br>
-read paired (0x1)<br>
-read mapped in proper pair (0x2)<br>
-mate reverse strand (0x20)<br>
-first in pair (0x40)<br>
